File: src/easy/has-cycle.ts
import type { ListNode } from "./list-node";

export const hasCycle = <T>(head: ListNode<T> | null): boolean => {
	let slow = head;
	let fast = head;
	while (fast !== null && fast.next !== null && slow !== null) {
		slow = slow?.next ?? null;
		fast = fast.next.next;
		if (slow === fast) return true;
	}

	return false;
};

File: lib/easy/has_cycle.dart
import 'package:dark_coding_challenges/easy/merge_two_linked_lists.dart';

bool hasCycle(ListNode? head) {
  ListNode? fast = head;
  ListNode? slow = head;
  while (fast != null && fast.next != null) {
    slow = slow?.next;
    fast = fast.next?.next;
    if (fast == slow) return true;
  }
  return false;
}
